internal/provider/tar_archiver_test.go

```diff
@@ -0,0 +1,435 @@
+	}
+}
+
+func TestTarArchiver_Dir_Exclude_DoNotExcludeSymlinkDirectories(t *testing.T) {
+	tarFilePath := filepath.Join(t.TempDir(), "archive-dir-with-symlink-dir.tar.gz")
+
+	archiver := NewTarGzArchiver(tarFilePath)
+	if err := archiver.ArchiveDir("./test-fixtures", ArchiveDirOpts{
+		Excludes: []string{
+			"test-symlink-dir/file1.txt",
+			"test-symlink-dir-with-symlink-file/test-symlink.txt",
+		},
+	}); err != nil {
+		t.Fatalf("unexpected error: %s", err)
+	}
+
+	ensureTarContents(t, tarFilePath, map[string][]byte{
+		"test-dir/test-dir1/file1.txt":                         []byte("This is file 1"),
+		"test-dir/test-dir1/file2.txt":                         []byte("This is file 2"),
+		"test-dir/test-dir1/file3.txt":                         []byte("This is file 3"),
+		"test-dir/test-dir2/file1.txt":                         []byte("This is file 1"),
+		"test-dir/test-dir2/file2.txt":                         []byte("This is file 2"),
+		"test-dir/test-dir2/file3.txt":                         []byte("This is file 3"),
+		"test-dir/test-file.txt":                               []byte("This is test content"),
+		"test-dir-with-symlink-dir/test-symlink-dir/file1.txt": []byte("This is file 1"),
+		"test-dir-with-symlink-dir/test-symlink-dir/file2.txt": []byte("This is file 2"),
+		"test-dir-with-symlink-dir/test-symlink-dir/file3.txt": []byte("This is file 3"),
+		"test-dir-with-symlink-file/test-file.txt":             []byte("This is test content"),
+		"test-dir-with-symlink-file/test-symlink.txt":          []byte("This is test content"),
+		"test-symlink-dir/file2.txt":                           []byte("This is file 2"),
+		"test-symlink-dir/file3.txt":                           []byte("This is file 3"),
+		"test-symlink-dir-with-symlink-file/test-file.txt":     []byte("This is test content"),
+	})
+}
+
+func TestTarArchiver_Dir_Exclude_ExcludeSymlinkDirectories(t *testing.T) {
+	tarFilePath := filepath.Join(t.TempDir(), "archive-dir-with-symlink-dir.tar.gz")
+
+	archiver := NewTarGzArchiver(tarFilePath)
+	err := archiver.ArchiveDir("./test-fixtures", ArchiveDirOpts{
+		Excludes: []string{
+			"test-dir/test-dir1/file1.txt",
+			"test-symlink-dir-with-symlink-file/test-symlink.txt",
+		},
+		ExcludeSymlinkDirectories: true,
+	})
+
+	if err != nil {
+		t.Errorf("expected no error: %s", err)
+	}
+}
+
+func ensureTarContents(t *testing.T, tarfilepath string, wants map[string][]byte) {
+	t.Helper()
+
+	f, err := os.Open(tarfilepath)
+	if err != nil {
+		t.Fatalf("could not open tar.gz file: %s", err)
+	}
+	defer f.Close()
+
+	gzf, err := gzip.NewReader(f)
+	if err != nil {
+		t.Fatalf("could not open tar.gz file: %s", err)
+	}
+	defer gzf.Close()
+
+	tarReader := tar.NewReader(gzf)
+
+	i := 0
+	for {
+		header, err := tarReader.Next()
+
+		if err == io.EOF {
+			break
+		}
+
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		if len(wants) < i+1 {
+			t.Fatalf("unexpect file count in tar. expect %d", len(wants))
+		}
```

Comment:
Nit: This block could probably be moved to outside of the loop, as we only care about the final count of the tar files being equal to the length of `wants`.